connection_producer.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package snowflake
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"database/sql"
+	"encoding/pem"
+	"fmt"
+	"os"
+	"regexp"
+	"sync"
+
+	"github.com/hashicorp/vault/sdk/database/helper/connutil"
+	"github.com/mitchellh/mapstructure"
+	"github.com/snowflakedb/gosnowflake"
+)
+
+var (
+	ErrInvalidSnowflakeURL           = fmt.Errorf("invalid connection URL format, expect <account_name>.snowflakecomputing.com/<db_name>")
+	accountAndDBNameFromConnURLRegex = regexp.MustCompile(`^(.+)\.snowflakecomputing.com/(.+)$`) // Expected format: <account_name>.snowflakecomputing.com/<db_name>
+)
+
+type snowflakeConnectionProducer struct {
+	ConnectionURL         string `json:"connection_url"`
+	MaxOpenConnections    int    `json:"max_open_connections"`
+	MaxIdleConnections    int    `json:"max_idle_connections"`
+	MaxConnectionLifetime string `json:"max_connection_lifetime"`
+	Username              string `json:"username"`
+	Password              string `json:"password"`
+	PrivateKey            string `json:"private_key"`
+	UsernameTemplate      string `json:"username_template"`
+	DisableEscaping       bool   `json:"disable_escaping"`
+
+	Initialized bool
+	RawConfig   map[string]any
+	Type        string
+	snowflakeDB *sql.DB
+	sync.Mutex
+}
+
+func (c *snowflakeConnectionProducer) secretValues() map[string]string {
+	return map[string]string{
+		c.Password: "[password]",
+	}
+}
+
+func (c *snowflakeConnectionProducer) Init(ctx context.Context, initConfig map[string]interface{}, verifyConnection bool) (saveConfig map[string]interface{}, err error) {
+	c.Lock()
+	defer c.Unlock()
+
+	c.RawConfig = initConfig
+
+	decoderConfig := &mapstructure.DecoderConfig{
+		Result:           c,
+		WeaklyTypedInput: true,
+		TagName:          "json",
+	}
+
+	decoder, err := mapstructure.NewDecoder(decoderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	err = decoder.Decode(initConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(c.Password) > 0 {
+		// Return an error here once Snowflake ends support for password auth.
+	}
+
+	c.Initialized = true
+
+	if verifyConnection {
+		if _, err := c.Connection(ctx); err != nil {
+			c.close()
+			return nil, fmt.Errorf("error verifying connection: %w", err)
+		}
+	}
+
+	return initConfig, nil
+}
+
+func (c *snowflakeConnectionProducer) Initialize(ctx context.Context, config map[string]any, verifyConnection bool) error {
+	_, err := c.Init(ctx, config, verifyConnection)
+	return err
+}
+
+func (c *snowflakeConnectionProducer) Connection(ctx context.Context) (interface{}, error) {
+	// This is intentionally not grabbing the lock since the calling functions (e.g. CreateUser)
+	// are claiming it. (The locking patterns could be refactored to be more consistent/clear.)
```

Comment:
Suggest to remove this. Once we make the changes I suggested to use a single lock, the patterns should be consistent. Arguably, they could be more clear but I don't think this comment is useful.
```suggestion
	// are claiming it.
```